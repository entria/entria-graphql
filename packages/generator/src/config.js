/* @flow */
import path from 'path';
import merge from 'lodash.merge';
import fs from 'fs';
import pkgDir from 'pkg-dir';

const rootPath = pkgDir.sync('.') || '.';

let cacheConfig = null;

/**
 * Parse `.graphqlrc` config file and retrieve its contents
 * @param filePath {string} The path of the config file
 * @returns {*}
 */
const parseConfigFile = (filePath: string) => {
  const config = JSON.parse(fs.readFileSync(filePath, 'utf8'));

  const directories = Object.keys(config.directories).reduce((data, directory) => {
    if (directory === DIRECTORY_TYPE.SOURCE) {
      return {
        ...data,
        [directory]: `${rootPath}/${config.directories[directory]}`,
        [DIRECTORY_TYPE.SRC]: config.directories.source,
      };
    }

    return {
      ...data,
      [directory]: `${config.directories.source}/${config.directories.module}/${config.directories[directory]}`,
    };
  }, {});

  return {
    ...config,
    directories: {
      ...config.directories,
      ...directories,
    },
  };
};

export const DIRECTORY_TYPE = {
  SOURCE: 'source',
  SRC: 'src',
  MODULE: 'module',
  CONNECTION: 'connection',
  LOADER: 'loader',
  MODEL: 'model',
  MUTATION: 'mutation',
  MUTATION_TEST: 'mutation_test',
  TYPE: 'type',
  TYPE_TEST: 'type_test',
};

type DirectoryType = $Values<typeof DIRECTORY_TYPE>;
type Directories = {
  source: string,
  module: string,
  connection: string,
  loader: string,
  model: string,
  mutation: string,
  mutation_test: string,
  type: string,
  type_test: string,
  interface: string,
};

type Files = {
  schema: string,
};

type Config = {
  directories: Directories,
  files: Files,
}
/**
 * Get the `.graphqlrc` config file
 * @returns {object} The content of the config
 */
export const getCreateGraphQLConfig = (): Config => {
  // if (cacheConfig) return cacheConfig;

  // Use default config
  const defaultFilePath = path.resolve(`${__dirname}/graphqlrc.json`);

  const config = parseConfigFile(defaultFilePath);

  try {
    // Check if there is a `.graphqlrc` file in the root path
    const customConfig = parseConfigFile(`${rootPath}/.graphqlrc`);

    merge(config, customConfig);

    cacheConfig = config;

    // If it does, extend default config with it, so if the custom config has a missing line
    // it won't throw errors
    return config;
  } catch (err) {
    cacheConfig = config;
    // Return the default config if the custom doesn't exist
    return config;
  }
};

/**
 * Get a directory from the configuration file
 * @param directory {string} The name of the directory, e.g. 'source'/'mutation'
 * @returns {string} The directory path
 */
export const getConfigDir = (directory: DirectoryType) => getCreateGraphQLConfig().directories[directory];
