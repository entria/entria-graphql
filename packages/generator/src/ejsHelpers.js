/* @flow */
/**
 * Uppercase the first letter of a text
 * @param text {string}
 * @returns {string}
 */
export const uppercaseFirstLetter = (text: string) => `${text.charAt(0).toUpperCase()}${text.slice(1)}`;
/**
 * Camel cases text
 * @param text {string} Text to be camel-cased
 * @returns {string} Camel-cased text
 */
export const camelCaseText = (text: string) =>
  text.replace(/(?:^\w|[A-Z]|\b\w|\s+)/g, (match, index) => {
    if (+match === 0) {
      return '';
    }

    return index === 0 ? match.toLowerCase() : match.toUpperCase();
  });
