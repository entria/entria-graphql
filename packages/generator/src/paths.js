/* @flow */
import path from 'path';

export const getModulePath = (source: string, moduleName: string) => {
  return path.join(source, moduleName);
};

export const getTestPath = (source: string) => {
  return path.join(source, '__tests__');
};

export const getMutationPath = (source: string) => {
  return path.join(source, 'mutation');
};
