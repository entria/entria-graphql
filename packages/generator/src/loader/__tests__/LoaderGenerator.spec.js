import helper from 'yeoman-test';
import assert from 'yeoman-assert';
import path from 'path';

import {
  copyFixturesToModules,
  getFileContent,
} from '../../../test/helpers';
import { getConfigDir } from '../../config';
import { uppercaseFirstLetter } from '../../ejsHelpers';
import { getModulePath } from '../../paths';

const loaderGenerator = path.join(__dirname, '..');

it('generate a loader', async () => {
  const moduleName = 'example';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(loaderGenerator)
    .withArguments(name)
    .toPromise();

  const destinationDir = getConfigDir('loader');

  const modulePath = getModulePath(destinationDir, moduleName);
  const loaderFilepath = path.join(modulePath, `${name}Loader.js`);

  assert.file([
    loaderFilepath,
  ]);

  const files = {
    loader: getFileContent(path.join(folder, loaderFilepath)),
  };

  expect(files).toMatchSnapshot();
});

it('generate a loader with schema', async () => {
  const moduleName = 'post';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(loaderGenerator)
    .inTmpDir(dir => copyFixturesToModules(dir, moduleName))
    .withArguments('Post Post')
    .toPromise();

  const destinationDir = getConfigDir('loader');

  const modulePath = getModulePath(destinationDir, moduleName);
  const loaderFilepath = path.join(modulePath, `${name}Loader.js`);

  assert.file([
    loaderFilepath,
  ]);

  const files = {
    loader: getFileContent(path.join(folder, loaderFilepath)),
  };

  expect(files).toMatchSnapshot();
});

it('generate a loader with schema and without timestamps', async () => {
  const moduleName = 'user';
  const name = uppercaseFirstLetter(moduleName);

  const folder = await helper.run(loaderGenerator)
    .inTmpDir(dir => copyFixturesToModules(dir, moduleName))
    .withArguments('User User')
    .toPromise();

  const destinationDir = getConfigDir('loader');

  const modulePath = getModulePath(destinationDir, moduleName);
  const loaderFilepath = path.join(modulePath, `${name}Loader.js`);

  assert.file([
    loaderFilepath,
  ]);

  const files = {
    loader: getFileContent(path.join(folder, loaderFilepath)),
  };

  expect(files).toMatchSnapshot();
});
