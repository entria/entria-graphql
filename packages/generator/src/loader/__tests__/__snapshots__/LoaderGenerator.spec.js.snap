// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generate a loader 1`] = `
Object {
  "loader": "// @flow
import DataLoader from 'dataloader';
import { connectionFromMongoCursor, mongooseLoader } from '@entria/graphql-mongoose-loader';
import type { ConnectionArguments } from 'graphql-relay';

import ExampleModel from './ExampleModel';
import type { GraphQLContext } from '../../TypeDefinition';

type ExampleType = {
  id: string,
  _id: string,
  exampleField: string,
}

export default class Example {
  id: string;
  _id: string;
  exampleField: string;

  constructor(data: ExampleType) {
    this.id = data.id;
    this._id = data._id;
    this.exampleField = data.exampleField;
  }
}

export const getLoader = () => new DataLoader(ids => mongooseLoader(ExampleModel, ids));

const viewerCanSee = () => true;

export const load = async ({ dataloaders }: GraphQLContext, id: ?string) => {
  if (!id) return null;

  try {
    const data = await dataloaders.ExampleLoader.load(id.toString());

    if (!data) return null;

    return viewerCanSee() ? new Example(data) : null;
  } catch (err) {
    return null;
  }
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) => {
  return dataloaders.ExampleLoader.clear(id.toString());
};

export const loadExamples = async (context: GraphQLContext, args: ConnectionArguments) => {
  // TODO: specify conditions
  const examples = ExampleModel.find({});

  return connectionFromMongoCursor({
    cursor: examples,
    context,
    args,
    loader: load,
  });
};
",
}
`;

exports[`generate a loader with schema 1`] = `
Object {
  "loader": "// @flow
import DataLoader from 'dataloader';
import { connectionFromMongoCursor, mongooseLoader } from '@entria/graphql-mongoose-loader';
import type { ConnectionArguments } from 'graphql-relay';

import PostModel from './PostModel';
import type { GraphQLContext } from '../../TypeDefinition';

type PostType = {
  id: string,
  _id: string,
  title: string,
  author: string,
  slug: string,
  tags: string[],
  oldSlugs: string[],
  comments: string[],
  externalComments: string[],
  createdAt: Date,
  updatedAt: Date,
}

export default class Post {
  id: string;
  _id: string;
  title: string;
  author: string;
  slug: string;
  tags: string[];
  oldSlugs: string[];
  comments: string[];
  externalComments: string[];
  createdAt: Date;
  updatedAt: Date;

  constructor(data: PostType) {
    this.id = data.id;
    this._id = data._id;
    this.title = data.title;
    this.author = data.author;
    this.slug = data.slug;
    this.tags = data.tags;
    this.oldSlugs = data.oldSlugs;
    this.comments = data.comments;
    this.externalComments = data.externalComments;
    this.createdAt = data.createdAt;
    this.updatedAt = data.updatedAt;
  }
}

export const getLoader = () => new DataLoader(ids => mongooseLoader(PostModel, ids));

const viewerCanSee = () => true;

export const load = async ({ dataloaders }: GraphQLContext, id: ?string) => {
  if (!id) return null;

  try {
    const data = await dataloaders.PostLoader.load(id.toString());

    if (!data) return null;

    return viewerCanSee() ? new Post(data) : null;
  } catch (err) {
    return null
  };
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) => {
  return dataloaders.PostLoader.clear(id.toString());
};

export const loadPosts = async (context: GraphQLContext, args: ConnectionArguments) => {
  // TODO: specify conditions
  const posts = PostModel.find({});

  return connectionFromMongoCursor({
    cursor: posts,
    context,
    args,
    loader: load,
  });
};
",
}
`;

exports[`generate a loader with schema and without timestamps 1`] = `
Object {
  "loader": "// @flow
import DataLoader from 'dataloader';
import { connectionFromMongoCursor, mongooseLoader } from '@entria/graphql-mongoose-loader';
import type { ConnectionArguments } from 'graphql-relay';

import UserModel from './UserModel';
import type { GraphQLContext } from '../../TypeDefinition';

type UserType = {
  id: string,
  _id: string,
  name: string,
  password: string,
  email: string,
  active: boolean,
  lastLoginAt: Date,
}

export default class User {
  id: string;
  _id: string;
  name: string;
  password: string;
  email: string;
  active: boolean;
  lastLoginAt: Date;

  constructor(data: UserType) {
    this.id = data.id;
    this._id = data._id;
    this.name = data.name;
    this.password = data.password;
    this.email = data.email;
    this.active = data.active;
    this.lastLoginAt = data.lastLoginAt;
  }
}

export const getLoader = () => new DataLoader(ids => mongooseLoader(UserModel, ids));

const viewerCanSee = () => true;

export const load = async ({ dataloaders }: GraphQLContext, id: ?string) => {
  if (!id) return null;

  try {
    const data = await dataloaders.UserLoader.load(id.toString());

    if (!data) return null;

    return viewerCanSee() ? new User(data) : null;
  } catch (err) {
    return null
  };
};

export const clearCache = ({ dataloaders }: GraphQLContext, id: string) => {
  return dataloaders.UserLoader.clear(id.toString());
};

export const loadUsers = async (context: GraphQLContext, args: ConnectionArguments) => {
  // TODO: specify conditions
  const users = UserModel.find({});

  return connectionFromMongoCursor({
    cursor: users,
    context,
    args,
    loader: load,
  });
};
",
}
`;
